src/codecs/csl/index.ts

```diff
@@ -0,0 +1,228 @@
+/**
+ * Codec for CSL JSON
+ *
+ * Also acts a base codec for other bibliographic formats
+ * e.g. BibTeX. Based on https://citation.js.org/
+ */
+
+import * as stencila from '@stencila/schema'
+import { isA, isCreativeWork } from '@stencila/schema/dist/util'
+// @ts-ignore
+import Cite from 'citation-js'
+import Csl from 'csl-json'
+import { Encode, EncodeOptions, load } from '../..'
+import * as vfile from '../../util/vfile'
+import { errorNodeType, warnLossIfAny, warnLoss } from '../../log'
+
+export const mediaTypes = ['application/vnd.citationstyles.csl+json']
+
+export const extNames = ['csl']
+
+/**
+ * Parse CSL-JSON or other bibliographic format to a `Node`
+ */
+export async function decode(
+  file: vfile.VFile,
+  format: string = '@csl/object'
+): Promise<stencila.Node> {
+  const content: string = await vfile.dump(file)
+
+  let csls
+  try {
+    csls = await Cite.inputAsync(content, { forceType: format })
+  } catch (error) {
+    throw new Error(
+      `Error when parsing content of format ${format}: ${error.message}`
+    )
+  }
+
+  // TODO: work out what to return when more than one work e.g. a bibtex file
+  const csl = csls[0]
+
+  return decodeCsl(csl)
+}
+
+/**
+ * Encode a `Node` to CSL-JSON or other bibliographic format
+ *
+ * See https://citation.js.org/api/tutorial-output_options.html
+ * for formats and other options which could be used.
+ */
+export const encode: Encode = async (
+  node: stencila.Node,
+  options: EncodeOptions = {}
+): Promise<vfile.VFile> => {
```

Comment:
In the top level `encode` function we check the node type using type guards and either wrap it as needed e.g. `InlineContent` into a `CreativeWork` or log an `errorNodeType`